File: suc/swift/017.swift

import SwiftUI
import UIKit // Required for UIAppearance customizations

// MARK: - Color Extension for Hex Initialization
// Provides a convenient way to initialize SwiftUI Colors from hex values,
// mirroring the Android Color(0xFFRRGGBB) syntax.
extension Color {
    init(hex: UInt, alpha: Double = 1.0) {
        self.init(
            .sRGB, // Use sRGB color space
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: alpha
        )
    }
}

// MARK: - AppTokens
// Centralized definitions for colors, typography, shapes, and spacing,
// mirroring the Kotlin AppTokens object for easy modification and consistency.
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF00FF66)
        static let secondary = Color(hex: 0xFFCCCCCC)
        static let tertiary = Color(hex: 0xFF999999)
        static let background = Color(hex: 0xFF000000)
        static let surface = Color(hex: 0xFF0A0A0A)
        static let surfaceVariant = Color(hex: 0xFF141414)
        static let outline = Color(hex: 0xFF2A2A2A)
        static let success = Color(hex: 0xFF22C55E)
        static let warning = Color(hex: 0xFFF59E0B)
        static let error = Color(hex: 0xFFEF4444)
        static let onPrimary = Color(hex: 0xFF000000)
        static let onSecondary = Color(hex: 0xFFFFFFFF)
        static let onTertiary = Color(hex: 0xFFFFFFFF)
        static let onBackground = Color(hex: 0xFFE0E0E0)
        static let onSurface = Color(hex: 0xFFE0E0E0)
    }

    struct TypographyTokens {
        // Mapping Kotlin's TextStyle to SwiftUI's Font.
        // fontSize (sp) maps directly to SwiftUI's point size.
        // FontWeight.Bold -> .bold, FontWeight.Medium -> .medium, FontWeight.Normal -> .regular
        static let display = Font.system(size: 26, weight: .bold)
        static let title = Font.system(size: 16, weight: .medium)
        static let body = Font.system(size: 13, weight: .regular)
        static let label = Font.system(size: 11, weight: .medium)
    }

    struct Shapes {
        // RoundedCornerShape(X.dp) maps to CGFloat cornerRadius values.
        static let small: CGFloat = 6
        static let medium: CGFloat = 10
        static let large: CGFloat = 14
    }

    struct Spacing {
        // Dp values map directly to CGFloat for consistent spacing.
        static let xs: CGFloat = 4
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }
    // ShadowSpec and ElevationMapping are defined in Kotlin but not used in the UI,
    // so they are omitted here for brevity.
}

// MARK: - MixerSlider Component
// Reusable component for the individual audio mixer sliders.
struct MixerSlider: View {
    let label: String
    @Binding var value: Float // Use @Binding to allow external modification of the slider value

    var body: some View {
        VStack(alignment: .leading, spacing: AppTokens.Spacing.xs) {
            Text("\(label) \(Int(value * 100))%")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface)
            Slider(value: $value, in: 0...1)
                .frame(maxWidth: .infinity) // Equivalent to Modifier.fillMaxWidth()
                // Slider colors (thumb, active/inactive track) are customized globally
                // using UIAppearance in the App struct's init() to ensure atomic correspondence.
        }
    }
}

// MARK: - RootScreen
// The main UI screen, equivalent to the Composable RootScreen in Kotlin.
struct RootScreen: View {
    // @State variables to hold the mutable slider values, mirroring Kotlin's remember { mutableFloatStateOf(...) }
    @State private var bass: Float = 0.4
    @State private var mid: Float = 0.5
    @State private var treble: Float = 0.6
    @State private var master: Float = 0.7

    var body: some View {
        ZStack { // Use ZStack to layer the background behind the content.
            // Full screen background color, ignoring safe areas to extend to screen edges.
            AppTokens.Colors.background.ignoresSafeArea(.all)

            VStack(spacing: 0) { // Main content stack, no spacing between top bar and content.
                // Top Bar equivalent: Simulates CenterAlignedTopAppBar.
                // Material 3 CenterAlignedTopAppBar typically has a height of 64.dp.
                // The title text (26pt) is vertically centered within this fixed height.
                Text("Audio Mixer Terminal")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.primary)
                    .frame(maxWidth: .infinity) // Center aligned horizontally
                    .padding(.vertical, (64 - 26) / 2) // Calculate vertical padding to center 26pt text in 64pt height
                    .frame(height: 64) // Fixed height for the top bar
                    .background(AppTokens.Colors.background) // Ensure top bar background is black

                // Content area: Contains sliders, progress bar, and button.
                VStack(spacing: AppTokens.Spacing.lg) { // Spacing between elements in this column.
                    MixerSlider(label: "Bass", value: $bass)
                    MixerSlider(label: "Mid", value: $mid)
                    MixerSlider(label: "Treble", value: $treble)

                    // Spacer equivalent to Modifier.height(AppTokens.Spacing.lg)
                    Spacer().frame(height: AppTokens.Spacing.lg)

                    // LinearProgressIndicator equivalent
                    ProgressView(value: master)
                        .tint(AppTokens.Colors.primary) // Sets the active track color
                        .frame(maxWidth: .infinity) // Equivalent to Modifier.fillMaxWidth()
                        .frame(height: 6) // Fixed height 6.dp

                    Text("Master Volume \(Int(master * 100))%")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.onSurface)
                        .frame(maxWidth: .infinity, alignment: .leading) // Aligned to start

                    // Master volume Slider
                    Slider(value: $master, in: 0...1)
                        .frame(maxWidth: .infinity) // Equivalent to Modifier.fillMaxWidth()
                        // Slider colors customized globally via UIAppearance.

                    // Apply Settings Button
                    Button(action: {}) {
                        Text("Apply Settings")
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onPrimary)
                            .frame(maxWidth: .infinity) // Text fills button width
                            .padding(.vertical, (48 - 16) / 2) // Vertically center 16pt text in 48pt height
                    }
                    .background(AppTokens.Colors.primary) // Button background color
                    .cornerRadius(AppTokens.Shapes.medium) // Rounded corners (10.dp)
                    .frame(height: 48) // Fixed button height 48.dp
                }
                .padding(AppTokens.Spacing.lg) // Outer padding for the content column (16.dp on all sides)
                .frame(maxWidth: .infinity, maxHeight: .infinity) // Make content column fill remaining space
                .background(AppTokens.Colors.background) // Ensure content background is black
            }
        }
        .statusBarHidden(true) // Hides the top status bar, as required.
    }
}

// MARK: - App Entry Point
// The main application struct, conforming to the App protocol.
// This is the entry point for a SwiftUI application.
@main
struct AudioMixerApp: App {
    // Initialize UIAppearance for UIKit controls (UISlider, UIProgressView)
    // to match the atomic color requirements from the Android version.
    // These changes apply globally to all instances of these UIKit controls
    // used within the SwiftUI hierarchy.
    init() {
        // Customize UISlider appearance
        UISlider.appearance().minimumTrackTintColor = UIColor(AppTokens.Colors.primary) // Active track color
        UISlider.appearance().maximumTrackTintColor = UIColor(AppTokens.Colors.surfaceVariant) // Inactive track color
        UISlider.appearance().thumbTintColor = UIColor(AppTokens.Colors.primary) // Thumb color

        // Customize UIProgressView appearance
        UIProgressView.appearance().progressTintColor = UIColor(AppTokens.Colors.primary) // Filled progress color
        UIProgressView.appearance().trackTintColor = UIColor(AppTokens.Colors.surfaceVariant) // Unfilled track color
    }

    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}

// MARK: - Preview Provider
// Provides a preview of the RootScreen in Xcode's canvas.
struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: suc/kotlin/017.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "016AudioMixeren"
private const val UI_TYPE = "Media"
private const val STYLE_THEME = "Monochrome Terminal"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF00FF66)
        val secondary = Color(0xFFCCCCCC)
        val tertiary = Color(0xFF999999)
        val background = Color(0xFF000000)
        val surface = Color(0xFF0A0A0A)
        val surfaceVariant = Color(0xFF141414)
        val outline = Color(0xFF2A2A2A)
        val success = Color(0xFF22C55E)
        val warning = Color(0xFFF59E0B)
        val error = Color(0xFFEF4444)
        val onPrimary = Color(0xFF000000)
        val onSecondary = Color(0xFFFFFFFF)
        val onTertiary = Color(0xFFFFFFFF)
        val onBackground = Color(0xFFE0E0E0)
        val onSurface = Color(0xFFE0E0E0)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 13.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 11.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(14.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(0.dp, 0.dp, 0.dp, 0f)
        val level2 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
    }
}

private val AppColorScheme = darkColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    var bass by remember { mutableFloatStateOf(0.4f) }
    var mid by remember { mutableFloatStateOf(0.5f) }
    var treble by remember { mutableFloatStateOf(0.6f) }
    var master by remember { mutableFloatStateOf(0.7f) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = {
                    Text("Audio Mixer Terminal", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.primary)
                },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(AppTokens.Colors.background)
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg)
        ) {
            MixerSlider("Bass", bass) { bass = it }
            MixerSlider("Mid", mid) { mid = it }
            MixerSlider("Treble", treble) { treble = it }
            Spacer(Modifier.height(AppTokens.Spacing.lg))
            LinearProgressIndicator(
                progress = { master },
                color = AppTokens.Colors.primary,
                trackColor = AppTokens.Colors.surfaceVariant,
                modifier = Modifier.fillMaxWidth().height(6.dp)
            )
            Text(
                text = "Master Volume ${(master * 100).toInt()}%",
                style = MaterialTheme.typography.titleMedium,
                color = AppTokens.Colors.onSurface
            )
            Slider(
                value = master,
                onValueChange = { master = it },
                colors = SliderDefaults.colors(
                    thumbColor = AppTokens.Colors.primary,
                    activeTrackColor = AppTokens.Colors.primary,
                    inactiveTrackColor = AppTokens.Colors.surfaceVariant
                ),
                modifier = Modifier.fillMaxWidth()
            )
            Button(
                onClick = {},
                colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                shape = AppTokens.Shapes.medium,
                modifier = Modifier.fillMaxWidth().height(48.dp)
            ) {
                Text("Apply Settings", style = MaterialTheme.typography.titleMedium)
            }
        }
    }


}

@Composable
fun MixerSlider(label: String, value: Float, onChange: (Float) -> Unit) {
    Column(
        verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.xs),
        horizontalAlignment = Alignment.Start
    ) {
        Text("$label ${(value * 100).toInt()}%", style = MaterialTheme.typography.bodyMedium, color = AppTokens.Colors.onSurface)
        Slider(
            value = value,
            onValueChange = onChange,
            colors = SliderDefaults.colors(
                thumbColor = AppTokens.Colors.primary,
                activeTrackColor = AppTokens.Colors.primary,
                inactiveTrackColor = AppTokens.Colors.surfaceVariant
            ),
            modifier = Modifier.fillMaxWidth()
        )
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF000000)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}